rig/routing_table/ordered_covering.py

```diff
@@ -0,0 +1,752 @@
+    merge insertion position.
+
+    For example, in the (non-orthogonal) table::
+
+        00001 -> N S
+        00011 -> N S
+        00100 -> N S
+        00X00 -> N S
+        XX1XX -> 3 5
+
+    Merging the first four entries would generate the new key-mask ``00XXX``
+    which would be inserted above the entry with the key-mask ``XX1XX``.
+    However ``00XXX`` would stop the key ``00110`` from reaching its correct
+    route, that is ``00110`` would be covered by ``00XXX``. To avoid this one
+    could just abandon the merge entirely, but a better solution is to attempt
+    to reduce the merge such that it no longer covers any entries below it.
+
+    To do this we first identify the bits that ARE ``X`` s in the merged
+    key-mask but which are NOT ``X`` s in the entry that we're covering. For
+    this example this is the 3rd bit. We then look to remove from the merge any
+    entries which are either ``X`` s in this position OR have the same value as
+    in this bit as the aliased entry. As the 4th entry in the table has an
+    ``X`` in this position we remove it, and as the 3rd entry has a ``1`` we
+    also remove it.  For this example we would then consider merging only the
+    first two entries, leading to a new key-mask pair of ``000X1`` which can be
+    safely inserted between ``00X00`` and ``XX1XX``::
+
+        00100 -> N S
+        00X00 -> N S
+        000X1 -> N S
+        XX1XX -> 3 5
+
+    Returns
+    -------
+    :py:class:`~.Merge`
+        New merge with entries possibly removed. If the goodness of the merge
+        ever drops below `min_goodness` then an empty merge will be returned.
+    """
+    # Operation
+    # ---------
+    # While the merge is still better than `min_goodness` we determine which
+    # entries below it in the table it covers. For each of these covered
+    # entries we find which bits are Xs in the merged entry and are NOT Xs in
+    # the covered entry.
+    #
+    # For example:
+    #
+    #     Merged entry:      ...0XXX1...
+    #     Covered entry:     ...010XX...
+    #     Bits of interest:      ^^
+    #     Label used below:      mn
+    #
+    # In this case there are 2 ``settable`` bits of interest highlighted.
+    # Whenever we have multiple covered entries we care only about the entries
+    # with the fewest number of ``settable`` bits because these most constrain
+    # which entries we may remove from the merge to avoid covering up the lower
+    # entry.
+    #
+    # NOTE:
+    #   * If there is only 1 ``settable`` bit then we are very constrained in
+    #     terms of which entries must be removed from the merge to avoid
+    #     covering a lower entry.
+    #   * If there are no ``settable`` bits then we cannot possibly avoid
+    #     covering the lower entry - the only correct action is to return an
+    #     empty merge.
+    #
+    # Assuming that there were no covered entries with no ``settable`` bits
```

Comment:
"Assuming that there were no covered entries **without any** ``settable`` bits" ??